File: app/src/main/java/app/fit/fitndflow/domain/usecase/DeleteExerciseUseCase.java
package app.fit.fitndflow.domain.usecase;

import android.content.Context;

import java.util.List;

import app.fit.fitndflow.data.common.SharedPrefs;
import app.fit.fitndflow.domain.common.usecase.FitUseCase;
import app.fit.fitndflow.domain.model.CategoryModel;
import app.fit.fitndflow.domain.model.ExerciseModel;
import app.fit.fitndflow.domain.repository.FitnFlowRepository;
import io.reactivex.Single;

public class DeleteExerciseUseCase extends FitUseCase<Integer, List<ExerciseModel>> {

    private FitnFlowRepository fitnFlowRepository;
    private Context context;

    public DeleteExerciseUseCase(int inputParams, Context context, FitnFlowRepository fitnFlowRepository){
        super(inputParams);
        this.context = context;
        this.fitnFlowRepository = fitnFlowRepository;
    }

    @Override
    public Single<List<ExerciseModel>> buildUseCaseObservable() {
        return Single.create(emitter -> {
            try{
                String apiKey = SharedPrefs.getApikeyFromSharedPRefs(context);
                List<ExerciseModel> response = fitnFlowRepository.deleteExercise(inputParams, apiKey);
                emitter.onSuccess(response);
            } catch(Exception exception) {
                emitter.onError(exception);
            }
        });
    }
}

File: app/src/main/java/app/fit/fitndflow/domain/usecase/DeleteExerciseUseCase.kt
package app.fit.fitndflow.domain.usecase

import android.content.Context
import app.fit.fitndflow.data.common.SharedPrefs
import app.fit.fitndflow.domain.common.usecase.UseCase
import app.fit.fitndflow.domain.model.ExerciseModel
import app.fit.fitndflow.domain.repository.FitnFlowRepository
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow

class DeleteExerciseUseCase(val fitnFlowRepository: FitnFlowRepository, val context: Context) : UseCase<ExerciseToDeleteParams, List<ExerciseModel>>() {
    override fun run(params: ExerciseToDeleteParams): Flow<List<ExerciseModel>> = flow {
        val apiKey = SharedPrefs.getApikeyFromSharedPRefs(context)
        val exerciseToDelete = fitnFlowRepository.deleteExercise(params.exerciseId, apiKey)
        fitnFlowRepository.updateCurrentTrainingListCache(apiKey)
        emit(exerciseToDelete)
    }
}

data class ExerciseToDeleteParams(val exerciseId: Int)